### PR TITLE
Rate limiting.

### PR DESCRIPTION
**Pull request by peter-fogg**
# [ECOM-3896](https://openedx.atlassian.net/browse/ECOM-3896)

@clintonb @bderusha

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -40,6 +40,9 @@ clean: clean_static
 requirements:
 	pip install -qr requirements/local.txt --exists-action w
 
+local-requirements:
```
Comment (clintonb) [inline Makefile]: Revert.

File: course_discovery/apps/core/models.py [hunk 2]
```diff
@@ -25,3 +25,10 @@ class Meta(object):  # pylint:disable=missing-docstring
 
     def get_full_name(self):
         return self.full_name or super(User, self).get_full_name()
+
+
+class UserThrottleRate(models.Model):
+    """Model for configuring a rate limit per-user."""
+
+    user = models.ForeignKey(User)
+    rate = models.CharField(max_length=50)
```
Comment (clintonb) [inline course_discovery/apps/core/models.py]: Add help text clarifying the units of measurement.

Also, there should be validation on this field. See https://docs.djangoproject.com/en/1.8/ref/models/instances/#validating-objects. You'll need to override the form used by `ModelAdmin`.

File: Makefile (hunk 1 above)
Comment (peter-fogg) [inline Makefile]: I'm planning on reverting 166d592a38fc82086fe9449cdc2b108563f62c2f once this has passed review; I'm just keeping it on the branch to simplify my own development.

**Comment by clintonb** [comment]
It would be nice if staff/superusers were exempt from throttling.

**Comment by clintonb** [comment]
:+1:

**Comment by bderusha** [comment]
:+1: I do agree with clinton that we may want to add additional logic to exempt certain users from any limit.  But I don't think that should hold up this PR necessarily.

File: course_discovery/apps/core/models.py (hunk 2 above)
Comment (peter-fogg) [inline course_discovery/apps/core/models.py]: Will do.

**Comment by peter-fogg** [comment]
I like the idea of superusers being exempt from throttling -- I'll add that.

File: course_discovery/apps/core/models.py (hunk 2 above)
Comment (peter-fogg) [inline course_discovery/apps/core/models.py]: And done.

File: course_discovery/apps/core/forms.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+        cleaned_data = super(UserThrottleRateForm, self).clean()
+        rate = cleaned_data.get('rate')
+        if rate:
+            try:
+                num, period = rate.split('/')
+                int(num)  # Only evaluated for the (possible) side effect of a ValueError
+                if period not in ('second', 'minute', 'hour', 'day'):
+                    raise forms.ValidationError(_("period must be one of 'second', 'minute', 'hour', or 'day'."))
+            except ValueError:
+                raise forms.ValidationError(_("'rate' must be in the format defined by DRF."))
```
Comment (clintonb) [inline course_discovery/apps/core/forms.py]: Provide an example.

File: course_discovery/apps/core/forms.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+    def clean(self):
+        cleaned_data = super(UserThrottleRateForm, self).clean()
+        rate = cleaned_data.get('rate')
+        if rate:
+            try:
+                num, period = rate.split('/')
+                int(num)  # Only evaluated for the (possible) side effect of a ValueError
+                if period not in ('second', 'minute', 'hour', 'day'):
+                    raise forms.ValidationError(_("period must be one of 'second', 'minute', 'hour', or 'day'."))
```
Comment (clintonb) [inline course_discovery/apps/core/forms.py]: Minor: You can use the tuple above to generate this message.

File: course_discovery/apps/core/forms.py (hunk 3 above)
Comment (peter-fogg) [inline course_discovery/apps/core/forms.py]: Done.

File: course_discovery/apps/core/forms.py (hunk 4 above)
Comment (peter-fogg) [inline course_discovery/apps/core/forms.py]: Done.

**Comment by clintonb** [comment]
:+1:

File: Makefile (hunk 1 above)
Comment (clintonb) [inline Makefile]: Docker image updated: https://hub.docker.com/r/edxops/course-discovery/tags/.

**Pull request merged**